app/helpers/application_helper.rb

```diff
@@ -4,11 +4,10 @@ def invited?
   end
 
   def user_is_admin?
-    current_user.roles.where(name: 'admin').exists? if current_user
+    current_user.has_role?("admin") if current_user
```

Comment:
@srt32 My general impression is to always use AR b/c it is faster. But, this seems to be an example where a Ruby solution is faster. Why is that? What should I be on the lookout for to help me know when a Ruby solution is better?

Reply:
Great question. I'd typically start with AR queries and then move to an in memory implementation once you see a problem.

In this case, we saw a problem in that on the `user#index` action (which drives `/users` page) we would loop over each user and call `user_is_admin?` (and if that was false, then `user_is_staff?`), which is fine if we're calling it for one user because we do one query to ask the db "hey, does this user have this admin role?" but breaks down when we loop over all the users and ask that question because for each user (100's of them) we ask the database a question.

So, before this change, listing the users took these queries (more or less): 1) get all the users, please, 2) then for each user go get their roles. So, that's 1 query + (1 query per user). So, given 100 users, that's 101 queries (typically not good for a web request).

In this new implementation, we do: 1) get all the users, please, 2) get all the users' roles, please, and store them in memory and then the rest happens in memory. So, we have 2 total queries no matter how many users there are (a classic solution to an n+1 query).

The tradeoff here is that we're putting more records into memory but doing fewer database queries. In this case, this page is not paginated so eventually (in a few months maybe?) this page is going to explode in memory usage and we'll probably have to paginate or infinite scroll it but for now we are happy using more memory since it makes the page load quickly instead of timing out.

WDYT?

Reply:
Oh I completely forgot that .includes does eager loading! This makes total sense.